aea/protocols/generator.py

```diff
@@ -0,0 +1,505 @@
+    """Exception for parsing."""
+
+    pass
+
+
+class ProtocolTemplate:
+    """
+    Extract protocol specification data from YAML configuration file.
+
+    This class:
+    a) reads a protocol specification YAML file,
+    b) extracts the protocol specification data into a protocol template object,
+    c) performs consistency checking.
+    """
+
+    def __init__(self, config_address: str) -> None:
+        """
+        Instantiate a protocol template object.
+
+        :param config_address: the address of the protocol specification yaml file
+        :return: None
+        """
+        self.config_address = config_address
+        self.yaml_documents = list()  # type: List[Any]
+
+        self.name = ""
+        self.authors = ""
+        self.version = ""
+        self.license_config = ""
+        self.url = ""
+        self.description = ""
+
+        self.speech_acts = dict()  # type: Dict[str, List[Dict[str, str]]]
+
+    def load(self) -> bool:
+        """
+        Call the two methods that read and extract protocol specification Yaml file.
+
+        :return: None
+        """
+        try:
+            self._read_protocol_specification()
+            self._extract_specification()
+        except (yaml.YAMLError, ProtocolSpecificationParseError):
+            raise
+        return True
+
+    def _read_protocol_specification(self) -> None:
+        """
+        Read a protocol specification YAML file.
+
+        Open and read the specification YAML file, and extract the yaml documents.
+        Raise any exceptions caught during file handling and reading the yaml file.
+
+        :return: None
+        """
+        try:
+            with open(self.config_address, 'r') as stream:
+                yaml_data = yaml.safe_load_all(stream)
+                for document in yaml_data:
+                    self.yaml_documents.append(document)
+        except (OSError, IOError, yaml.MarkedYAMLError):
+            raise
+
+    def _extract_specification(self) -> None:
+        """
+        Extract protocol specification data.
+
+        Extract the protocol specification data form yaml document objects.
+        Raise protocol specification parsing exceptions if the specification's format is incorrect.
+
+        :return: None
+        """
+        if len(self.yaml_documents) >= 2 and self.yaml_documents[0] is not None and self.yaml_documents[1] is not None:
+            if "name" in self.yaml_documents[0]:
+                self.name = self.yaml_documents[0]["name"]
+                if type(self.name) is not str:
+                    raise ProtocolSpecificationParseError("Protocol's 'name' is not a string.")
+                if self.name == "":
+                    raise ProtocolSpecificationParseError("Protocol's 'name' is empty.")
+            else:
+                raise ProtocolSpecificationParseError("Protocol name could not be found.")
+            if "authors" in self.yaml_documents[0]:
```

Comment:
Seems you could cut this down drastically by looping over a Dict[str, Any] where the keys are authors, versions etc and the values their types.

Reply:
Done